### DISCUSSION TITLE
Custom server with dynamic routing ?

### DISCUSSION THREAD

**Discussion by msreekm** [vercel/next.js#14330] [Help] [answered]
How are folks doing dynamic routing with custom server, we are currently using friday/next-routes with useFileSystemPublicRoutes: false,
in next.config.js, but this library is not maintained for past 2 years and want to see what are the alternative options for custom server?

**Comment by jamesmosier** [comment, marked as answer]
I have this in an old project that uses a custom Express server and it allows for dynamic routes that point to a file in my Next project:

```js
const router = express.Router();
router.get('/dashboard/item/:id', (req, res) => {
  return app.render(req, res, '/dashboard/item/id', { // `pages/dashboard/item/id.js` is the name of the page in my directory
    ...req.params,
    ...req.query,
  });
});
```

Other than that I have been using Next's dynamic routes in conjunction with my custom server and that has worked out just fine!

**Reply by msreekm** [reply]
thanks ,i was wondering why the dodcumentation says that "**may result in the same content being served from multiple paths"**   

By default, Next will serve each file in the pages folder under a pathname matching the filename. If your project uses a custom server, this behavior may result in the same content being served from multiple paths, which can present problems with SEO and UX."  https://nextjs.org/docs/advanced-features/custom-server

**Reply by jamesmosier** [reply]
Because if you don't use `useFileSystemPublicRoutes: false` then in this example I would have 2 routes for the same file, `domain.com/dashboard/foobar/123` and `domain.com/dashboard/item/id`

```js
const router = express.Router();
router.get('/dashboard/foobar/:id', (req, res) => {
  return app.render(req, res, '/dashboard/item/id', {
    ...req.params,
    ...req.query,
  });
});
```

File structure:
```
pages/
  item/
    id.js

```

**Reply by gaurav5430** [reply]
Thanks. That clarifies a few things.

This also means if I am using a custom server but not remapping any file based routes,  it would be fine for me to have the flag set to true?

If I set the flag to false, would it now be my responsibility to make sure that routes are served appropriately from the file paths?

**Reply by jamesmosier** [reply]
Correct, you don't need to set this flag if you are using a custom server. The default in Next for `useFileSystemPublicRoutes` is `true`.

If you use `useFileSystemPublicRoutes: false` then you are responsible for pointing the routes (url) to the file in your project.

**Comment by ghost** [comment]
Can I use `getStaticProps` and `getStaticPaths` with Custom Server ?

**Reply by msreekm** [reply]
yes, should not be an issue,

**Reply by glassdimlygr** [reply]
...but it won't render the pages as html/static
